src/client/Client.cc

```diff
@@ -16370,14 +16382,28 @@ mds_rank_t Client::_get_random_up_mds() const
 {
   ceph_assert(ceph_mutex_is_locked_by_me(client_lock));
 
-  std::set<mds_rank_t> up;
+  std::set<mds_rank_t> up, stopping, valid;
   mdsmap->get_up_mds_set(up);
+  mdsmap->get_mds_set(stopping, MDSMap::STATE_STOPPING);
 
-  if (up.empty())
-    return MDS_RANK_NONE;
-  std::set<mds_rank_t>::const_iterator p = up.begin();
-  for (int n = rand() % up.size(); n; n--)
+  // Try to skip the stopping MDSs
+  std::set_difference(up.begin(), up.end(), stopping.begin(), stopping.end(),
+                      std::inserter(valid, valid.end()));
+  if (valid.empty()) {
+    if (stopping.empty())
+      return MDS_RANK_NONE;
+
+    auto p = stopping.begin();
+    for (int n = rand() % stopping.size(); n; n--) {
+      ++p;
+    }
+    return *p;
+  }
+
+  auto p = valid.begin();
+  for (int n = rand() % valid.size(); n; n--) {
```

Comment:
```suggestion
  p = std::advance(p, ceph::util::generate_random_number<uint64_t>(0, valid.size());
```

Might as well remove `rand()` which is a code smell these days.

Reply:
Please double check if the upper bound is inclusive.